### PR TITLE
LPS-198713 Move deleted KB articles & folders to Recycle Bin

### PR DESCRIPTION
**Pull request by marcogalluzzi**
https://liferay.atlassian.net/browse/LPS-198713

Enable feature flag: [LPS-188058](https://liferay.atlassian.net/browse/LPS-188058)

### CODE REVIEW

**Review by ambrinchaudhary — APPROVED** [review]
Frontend changes LGTM.

**Review by AliciaGarciaGarcia — CHANGES_REQUESTED** [review]
The main issue: I think that needs upgrade of tables

image


```
2023-11-08 09:23:00.197 ERROR [http-nio-8080-exec-7][SqlExceptionHelper:142] Unknown column 'status' in 'where clause'
2023-11-08 09:23:00.216 ERROR [http-nio-8080-exec-7][PortletRequestDispatcherImpl:295] Unable to dispatch request: com.liferay.portal.kernel.exception.SystemException: com.liferay.portal.kernel.dao.orm.ORMException: com.liferay.portal.kernel.dao.orm.ORMException: javax.persistence.PersistenceException: org.hibernate.exception.SQLGrammarException: could not extract ResultSet
2023-11-08 09:23:00.218 ERROR [http-nio-8080-exec-7][PortletServlet:109] Unable to process portlet com_liferay_knowledge_base_web_portlet_AdminPortlet: org.apache.jasper.JasperException: com.liferay.portal.kernel.exception.SystemException: com.liferay.portal.kernel.dao.orm.ORMException: com.liferay.portal.kernel.dao.orm.ORMException: javax.persistence.PersistenceException: org.hibernate.exception.SQLGrammarException: could not extract ResultSet
javax.portlet.PortletException: org.apache.jasper.JasperException: com.liferay.portal.kernel.exception.SystemException: com.liferay.portal.kernel.dao.orm.ORMException: com.liferay.portal.kernel.dao.orm.ORMException: javax.persistence.PersistenceException: org.hibernate.exception.SQLGrammarException: could not extract ResultSet
```

File: modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/portlet/action/DeleteKBArticlesAndFoldersMVCActionCommand.java [hunk 1]
```diff
@@ -36,20 +44,58 @@ protected void doProcessAction(
...
-		long[] deleteKBArticleResourcePrimKeys = ParamUtil.getLongValues(
+		String cmd = ParamUtil.getString(actionRequest, Constants.CMD);
+		long[] kbArticleResourcePrimKeys = ParamUtil.getLongValues(
 			actionRequest, "rowIdsKBArticle");
+		long[] kbFolderIds = ParamUtil.getLongValues(
+			actionRequest, "rowIdsKBFolder");
 
-		for (long deleteKBArticleResourcePrimKey :
-				deleteKBArticleResourcePrimKeys) {
+		if (cmd.equals(Constants.MOVE_TO_TRASH)) {
```
[the remark this answers is not in the document]
Comment (marcogalluzzi) [reply modules/apps/knowledge-base/knowledge-base-web/src/main/java/com/liferay/knowledge/base/web/internal/portlet/action/DeleteKBArticlesAndFoldersMVCActionCommand.java]: Suggestion added. See [c8d636c](https://github.com/liferay-lima/liferay-portal/pull/4876/commits/c8d636c8db0041311e7f1d77f597fe5ea5b8a819)

File: modules/apps/knowledge-base/knowledge-base-api/src/main/java/com/liferay/knowledge/base/model/KBFolderTable.java [hunk 2]
```diff
@@ -57,6 +57,14 @@ public class KBFolderTable extends BaseTable<KBFolderTable> {
 		"description", String.class, Types.VARCHAR, Column.FLAG_DEFAULT);
 	public final Column<KBFolderTable, Date> lastPublishDate = createColumn(
 		"lastPublishDate", Date.class, Types.TIMESTAMP, Column.FLAG_DEFAULT);
+	public final Column<KBFolderTable, Integer> status = createColumn(
```
[the remark this answers is not in the document]
Comment (marcogalluzzi) [reply modules/apps/knowledge-base/knowledge-base-api/src/main/java/com/liferay/knowledge/base/model/KBFolderTable.java:60]: Missing upgrade added. See [1b667f5](https://github.com/liferay-lima/liferay-portal/pull/4876/commits/1b667f580ea2822a35c3ac45ea710514f9368cd1).

**Comment by marcogalluzzi** [comment]
ci:test:sf

**Review by AliciaGarciaGarcia — CHANGES_REQUESTED** [review]
When you create a new folder, the values of the new fields (status) are incorrect.

**Comment by marcogalluzzi** [comment]
ci:test:sf

**Comment by marcogalluzzi** [comment]
> When you create a new folder, the values of the new fields (status) are incorrect.

Hi @AliciaGarciaGarcia I've added a new commit to correct this: [98fc7d3](https://github.com/liferay-lima/liferay-portal/pull/4876/commits/98fc7d3e9271e07ce1ccf85b9b1616630f70bb1a)

**Comment by AliciaGarciaGarcia** [comment]
> > When you create a new folder, the values of the new fields (status) are incorrect.
> 
> Hi @AliciaGarciaGarcia I've added a new commit to correct this: [98fc7d3](https://github.com/liferay-lima/liferay-portal/pull/4876/commits/98fc7d3e9271e07ce1ccf85b9b1616630f70bb1a)

you checked on the` public KBFolder updateKBFolder( `if is needed too?

**Comment by marcogalluzzi** [comment]
> > > When you create a new folder, the values of the new fields (status) are incorrect.
> > 
> > 
> > Hi @AliciaGarciaGarcia I've added a new commit to correct this: [98fc7d3](https://github.com/liferay-lima/liferay-portal/pull/4876/commits/98fc7d3e9271e07ce1ccf85b9b1616630f70bb1a)
> 
> Hi @marcogalluzzi have you checked on the`public KBFolder updateKBFolder(` if is this needed too?

Hi @AliciaGarciaGarcia , I think it is not needed there. The update of the `status` (and related fields) is done in public method `updateStatus`

**Comment by AliciaGarciaGarcia** [comment]
ci:forward

**Comment by WYZ-Issac** [comment]
ci:test:stable

**Comment by WYZ-Issac** [comment]
ci:stop

**Comment by WYZ-Issac** [comment]
ci:test:stable

**Comment by WYZ-Issac** [comment]
Hi @marcogalluzzi, these failures only occur in your commit. Could you please take a look? Thanks.
[PortalLogAssertorTest-modules-integration-mysql57-jdk8/0/0](https://testray.liferay.com/zh/home/-/testray/case_results/2457929480)
[com.liferay.knowledge.base.service.test.KBArticleLocalServiceTest](https://testray.liferay.com/zh/home/-/testray/case_results/2457929492)
[com.liferay.knowledge.base.service.test.KBFolderLocalServiceTest](https://testray.liferay.com/zh/home/-/testray/case_results/2457929496)

**Comment by marcogalluzzi** [comment]
ci:test:sf

**Comment by marcogalluzzi** [comment]
Thanks @WYZ-Issac for catching the errors, I could solve all of them.

@AliciaGarciaGarcia can you please review the last 3 commits? 
Please note that commit [cde5d46](https://github.com/liferay-lima/liferay-portal/pull/4876/commits/cde5d46dd3b1185b718001e7c38b66483c239033) solves the case where we want to get all KB folders and articles with status DRAFT, which is only supported by KB articles.

**Review by AliciaGarciaGarcia — APPROVED** [review]

**Comment by AliciaGarciaGarcia** [comment]
ci:forward

**Pull request closed without merging**